http/native_server.ts

```diff
@@ -0,0 +1,369 @@
+ *       ) ?? "Unknown"}`;
+ *
+ *       return new Response(body, { status: 200 });
+ *     });
+ *
+ * @param {Deno.Listener} listener The listener to accept connections from.
+ * @param {HttpHandler} handler The handler for individual HTTP requests.
+ */
+export async function serve(
+  listener: Deno.Listener,
+  handler: HttpHandler,
+): Promise<void> {
+  const server = new Server(listener);
+
+  for await (const requestEvent of server) {
+    requestEvent.respondWith(
+      handler(requestEvent.request, requestEvent.connInfo),
+    );
+  }
+}
+
+/**
+ * Create a listener on the given address, accept incoming connections, and
+ * handle requests on these connections with the given handler.
+ *
+ *     const options = { port: 8000 };
+ *
+ *     listenAndServe(options, (request) => {
+ *       const body = `Your user-agent is:\n\n${request.headers.get(
+ *         "user-agent",
+ *       ) ?? "Unknown"}`;
+ *
+ *       return new Response(body, { status: 200 });
+ *     });
+ *
+ * @param {AddrOptions} addr The TCP address to listen on.
+ * @param {HttpHandler} handler The handler for individual HTTP requests.
+ */
+export async function listenAndServe(
+  addr: AddrOptions,
+  handler: HttpHandler,
+): Promise<void> {
+  let listenOptions;
+
+  if (typeof addr === "undefined" || addr === null) {
+    listenOptions = { port: 80 };
+  } else if (typeof addr === "number") {
+    listenOptions = { port: addr };
+  } else if (typeof addr === "string") {
+    listenOptions = _parseAddrFromStr(addr);
+  } else {
+    listenOptions = addr;
+  }
+
+  const listener = Deno.listen(listenOptions);
+
+  await serve(listener, handler);
+}
+
+/**
+ * Create a listener on the given address, accept incoming connections, upgrade
+ * them to TLS, and handle requests on these connections with the given handler.
+ *
+ *     const options = { port: 8000 };
+ *     const certFile = "/path/to/localhost.crt"
+ *     const keyFile = "/path/to/localhost.key"
+ *
+ *     listenAndServeTls(options, certFile, keyFile, (request) => {
+ *       const body = `Your user-agent is:\n\n${request.headers.get(
+ *         "user-agent",
+ *       ) ?? "Unknown"}`;
+ *
+ *       return new Response(body, { status: 200 });
+ *     });
+ *
+ * @param {AddrOptions} addr The TCP address to listen on.
+ * @param {string} certFile The path to the file containing the TLS certificate.
+ * @param {string} keyFile The path to the file containing the TLS private key.
+ * @param {HttpHandler} handler The handler for individual HTTP requests.
+ */
+export async function listenAndServeTls(
+  addr: AddrOptions,
+  certFile: string,
+  keyFile: string,
+  handler: HttpHandler,
+): Promise<void> {
+  let listenOptions;
+
+  if (typeof addr === "undefined" || addr === null) {
+    listenOptions = { port: 80 };
+  } else if (typeof addr === "number") {
+    listenOptions = { port: addr };
+  } else if (typeof addr === "string") {
+    listenOptions = _parseAddrFromStr(addr);
```

Comment:
This returns a default port of `80` whereas suspect better default for tls is `443`.

Reply:
c1ba097